Fix ActionListener.map exception handling

henningandersen opened the pull request:
map would call listener.onFailure for exceptions from
listener.onResponse, but this means we could double trigger some
listeners which is generally unexpected. Instead, we should assume that
a listener's onResponse (and onFailure) implementation is responsible
for its own exception handling.

This affects many APIs across the code base. This is the first of a series
of PRs changing exception handling to adhere to the principle that an
ActionListener implementation is responsible for its own exception handling.

The demo program test [here](https://github.com/henningandersen/elasticsearch/blob/demonstrate_exception_handling/server/src/test/java/org/elasticsearch/action/ExceptionHandlingIT.java) illustrates the current inconsistencies in how exceptions in listeners are handled.

henningandersen committed:
Fix ActionListener.map exception handling

map would call listener.onFailure for exceptions from
listener.onResponse, but this means we could double trigger some
listeners which is generally unexpected. Instead, we should assume that
a listener's onResponse (and onFailure) implementation is responsible
for its own exception handling.

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Core)

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Distributed)

henningandersen committed:
Fix checkstyle

pugnascotia reviewed (approved):
Changes look good.

How do you feel about adding some JavaDoc to the new test cases, to explain what the expectations are?

original-brownbear commented on server/src/main/java/org/elasticsearch/action/ActionListener.java:
```diff
@@ -136,14 +136,27 @@ public void onFailure(Exception e) {
+                delegate.onFailure(e);
+                return;
+            }
+            delegate.onResponse(mapped);
```
I see the point in this change, but note that I added the `.map` shortcut back when I added it to dry up a bunch of `ActionListener.wrap(..., listener::onFailure)` spots.
I think we'd basically have to audit every spot that we use `.map` in now and make sure that the `listener`/`delegate` will actually handle it's own `onResponse` failures (from a quick read over the spots we use `map` in this may already hold true).

Maybe we should `assert` this and do something like:

```java
try {
   delegate.onResponse(mapped);
} catch (Exception e) {
   assert false: e;
}
```

henningandersen replied:
I added the assert. Local CI seems happy about it.

henningandersen committed:
Add javadoc

henningandersen commented:
Thanks @pugnascotia , I added javadoc here: 198651b22672b353955779eb074074686e46e7e5

henningandersen committed:
Try assert

henningandersen committed:
Fix SearchTransportService DFS action

The DFS action relied on map notifying onFailure (sort of, at least this
way it is bwc). But there seems to be no reason it cannot simply use
the ChannelActionListener, so change it into using that.

henningandersen committed:
Adapt to asserting on failure.

henningandersen committed:
Remove test that cannot live with assertion error.

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_map_exception_handling

henningandersen commented:
Failure already reported [here](https://github.com/elastic/elasticsearch/issues/49753#issuecomment-575598706).

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_map_exception_handling

henningandersen commented:
Removed WIP on this, since I think this can go in alone.

ywelsch commented on server/src/main/java/org/elasticsearch/action/ActionListener.java:
```diff
@@ -136,14 +136,32 @@ public void onFailure(Exception e) {
+            try {
+                delegate.onResponse(mapped);
+            } catch (RuntimeException e) {
+                assert false : new AssertionError("map: listener.onResponse failed", e);
```
why assert here but not when calling `delegate.onFailure(e);`?

henningandersen replied:
Thanks, added that in ca31964870ea47b1391fa0cae78ee3c9272a7dcd and tests seems unaffected.

ywelsch commented on server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java:
```diff
@@ -306,27 +305,11 @@ public static void registerRequestHandler(TransportService transportService, Sea
-            });
+            (request, channel, task) ->
+                searchService.executeDfsPhase(request, (SearchShardTask) task,
+                    new ChannelActionListener<>(channel, DFS_ACTION_NAME, request))
```
`ChannelActionListener` also has this weird double-sending logic.

henningandersen replied:
Yes, but the code used to propagate the exception out. This seemed to be a left-over from when the `ChannelActionListener` was introduced. The old `map` exception handling would ensure `onFailure` were called on exception. This means this change is effectively a no-op now (with the caveat that `onFailure` will be called on exceptions like for all other `ChannelActionListener` usages).

We should notice that `DirectTransportChannel` will not bubble out exceptions from invoking the `TransportResponseHandler`. So it seems likely that the primary exceptions bubbled out are related to communicating the response over a wire, in which case invoking `onFailure` might be desirable (for instance an NPE on serialization).

So I would like to keep using `ChannelActionListener` here and then deal with `ChannelActionListener` in a follow-up. WDYT?

ywelsch replied:
++

ywelsch commented on server/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexIT.java:
```diff
@@ -378,5 +378,4 @@ public void testIndexNameInResponse() {
 
         assertEquals("Should have index name in response", "foo", response.index());
     }
-
```
undo

ywelsch reviewed (commented):
I have not gone through all callers of this method to check whether this is safe, but left a few small comments here.

henningandersen committed:
Review comments

Also assert that onFailure does not fail.

original-brownbear commented:
@henningandersen see failure here https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-2/14206/testReport/junit/org.elasticsearch.license/MachineLearningLicensingTests/testAutoCloseJobWithDatafeed/ ... we tripped a new assertion I think.

henningandersen commented:
Thanks, Armin. I reported that build failure [here](https://github.com/elastic/elasticsearch/issues/51285).

henningandersen commented:
@elasticmachine run elasticsearch-ci/2

original-brownbear commented:
@elasticmachine run elasticsearch-ci/2 (same ML failure again)

Thanks Henning

original-brownbear commented on server/src/main/java/org/elasticsearch/action/ActionListener.java:
```diff
@@ -136,14 +136,49 @@ public void onFailure(Exception e) {
-     * @param listener Listener to delegate to
+     * Notice that it is considered a bug if the listener's onResponse or onFailure fails. onResponse failures will not call onFailure.
+     *
+     * If the function fails, the listener's onFailure handler will be called. The principle is that the code using this is responsible for
```
Maybe word this a little differently, this reads somewhat confusing (IMO):

```
The principle is that the mapped listener will handle exceptions from the mapping function {@code fn} but it is the responsibility of {@code delegate} to handle its own exceptions inside `onResponse` and `onFailure`.
```

?

henningandersen replied:
Thanks Armin, I used your improved comment here: 09b55769e9e3aab45789b019d362111f51fc8e35

original-brownbear reviewed (commented):
I'm totally on board with this (just one comment on the docs) but lets fix the failing ML spot first maybe and get a few green test runs in before we merge?

henningandersen committed:
Merge remote-tracking branch 'origin/master' into fix_map_exception_handling

henningandersen committed:
Armins improved javadoc

original-brownbear reviewed (approved):
LGTM thanks Henning :)

henningandersen commented:
@elasticmachine test this please

henningandersen commented:
Build failure looks unrelated, reported it here: #51347 .
@elasticmachine run elasticsearch-ci/2

henningandersen commented:
One more test round:
@elasticmachine test this please

henningandersen commented:
> I have not gone through all callers of this method to check whether this is safe

I went through all calls in production (not test) code. Most are "clearly" OK, falling into one of these categories:

1. Clearly not able to throw except in severe bug cases like NPE.
2. Inner listener will not throw since it is either ActionListener.wrap og ChannelActionListener - though there is an addendum to that statement below.
3. Guarded by assertions enabled check (i.e. not active in production).
4. A choice is made to use map or not, clearly the intention was not to get the exception behavior (e.g. `TransportBulkAction.BulkRequestModifier.wrapActionListenerIfNeeded`).
5. Ultimately caught and `onFailure` called (transport service, security, ActionRunnable, StepListener/ListenableFuture).
6. Used with NotifyOnceListener inside only.

Following I did not follow to the end:

ProxyConnectionStrategy.clusterNameValidator: its use of counting will clearly be wrong if onResponse fails and onFailure is then called, but the effect of the rest is hard to judge.

PersistentTasksService: this affects all methods here. I checked a subset of the usages only:
* startPersistentTask:
  * CCR/resume follow ends in CCR `ResponseHandler` which does counting that will disregard an additional `onFailure` call anyway.
  * Rollup and transform uses wrap to create the listener
* sendCompletionRequest: only has trace logging listeners.
* updatePersistentTaskState: has a variety of listeners sent to it. Many use `wrap`, but some are complex and hard to chase to the end (so I did not).

The small addendum mentioned previously is that I did not consider the effect of having for instance `map(map(wrap(onResponse, onFailure)))` where `onResponse` and `onFailure` are functions that both throw. In that case, each additional `map` wrapping would previously invoke `onFailure` once more, such that `onFailure` in this case would be called 3 times.

henningandersen commented:
Notice that this PR was merged, then reverted. After #51646 was merged, the commit was cherry-picked (identical, no changes) to reintroduce it.
